crates/bevy_reflect/src/set.rs

```diff
@@ -0,0 +1,555 @@
+    fn into_iter(self) -> Self::IntoIter {
+        self.values.into_iter()
+    }
+}
+
+impl<'a> ExactSizeIterator for SetIter<'a> {}
+
+/// Compares a [`Set`] with a [`Reflect`] value.
+///
+/// Returns true if and only if all of the following are true:
+/// - `b` is a set;
+/// - `b` is the same length as `a`;
+/// - For each value pair in `a`, `b` contains the value too,
+///   and [`Reflect::reflect_partial_eq`] returns `Some(true)` for the two values.
+///
+/// Returns [`None`] if the comparison couldn't even be performed.
+#[inline]
+pub fn set_partial_eq<M: Set>(a: &M, b: &dyn Reflect) -> Option<bool> {
+    let ReflectRef::Set(set) = b.reflect_ref() else {
+        return Some(false);
+    };
+
+    if a.len() != set.len() {
+        return Some(false);
+    }
+
+    for value in a.iter() {
+        if let Some(set_value) = set.get(value) {
+            let eq_result = value.reflect_partial_eq(set_value);
+            if let failed @ (Some(false) | None) = eq_result {
+                return failed;
+            }
+        } else {
+            return Some(false);
+        }
+    }
+
+    Some(true)
+}
+
+/// The default debug formatter for [`Set`] types.
+///
+/// # Example
+/// ```
+/// # use bevy_utils::HashSet;
+/// use bevy_reflect::Reflect;
+///
+/// let mut my_set = HashSet::new();
+/// my_set.insert(String::from("Hello"));
+/// println!("{:#?}", &my_set as &dyn Reflect);
+///
+/// // Output:
+///
+/// // {
+/// //   "Hello",
+/// // }
+/// ```
+#[inline]
+pub fn set_debug(dyn_set: &dyn Set, f: &mut Formatter<'_>) -> std::fmt::Result {
+    let mut debug = f.debug_set();
+    for value in dyn_set.iter() {
+        debug.entry(&value as &dyn Debug);
+    }
+    debug.finish()
+}
+
+/// Applies the elements of reflected set `b` to the corresponding elements of set `a`.
+///
+/// If a value from `b` does not exist in `a`, the value is cloned and inserted.
+///
+/// # Panics
+///
+/// This function panics if `b` is not a reflected set.
+#[inline]
+pub fn set_apply<M: Set>(a: &mut M, b: &dyn Reflect) {
+    if let ReflectRef::Set(set_value) = b.reflect_ref() {
+        for b_value in set_value.iter() {
+            if a.get(b_value).is_none() {
+                a.insert_boxed(b_value.clone_value());
+            }
+        }
+    } else {
+        panic!("Attempted to apply a non-set type to a set type.");
+    }
+}
+
+/// Tries to apply the elements of reflected set `b` to the corresponding elements of set `a`
+/// and returns a Result.
+///
+/// If a key from `b` does not exist in `a`, the value is cloned and inserted.
+///
+/// # Errors
+///
+/// This function returns an [`ApplyError::MismatchedKinds`] if `b` is not a reflected set or if
+/// applying elements to each other fails.
+#[inline]
+pub fn set_try_apply<S: Set>(a: &mut S, b: &dyn Reflect) -> Result<(), ApplyError> {
+    if let ReflectRef::Set(set_value) = b.reflect_ref() {
+        for b_value in set_value.iter() {
+            if let Some(a_value) = a.get_mut(b_value) {
```

Comment:
I'm a bit surprised that this works differently from `set_apply` in the case that `a` contains a matching entry. Is this intentional?

Reply:
Good catch actually 🎯 I think this is because I only recently added the `get_mut` method to the trait which is kind of required. I changed `set_apply` to look more like `set_try_apply` (without the "try")